### DISCUSSION TITLE
Is there softSPI available on pyb?

### DISCUSSION THREAD

**Discussion by Meekdai** [micropython/micropython#13179] [STM32 / Pyboard]
I have an application that needs to use multi-channel SPI. The micropython version I am using is v1.11, so `machine.SoftSPI` cannot be used.
Is there softSPI available on pyb? If not, is it possible to port the Arduino library ([SoftSPI](https://github.com/MajenkoLibraries/SoftSPI)) to micropython?

**Comment by robert-hh** [comment]
The STM32 port for PyBoard support machine.SoftSPI, like most of the MicroPython ports.

**Reply by Meekdai** [reply]
Unfortunately, the v1.11 version does not yet have the function of softSPI. I don't want to update the version because a lot of the code in the project is based on v1.11. It would be a lot of work to update the version, so I wonder if someone has implemented an independent library.

```python
>>> import machine
>>> machine.SoftSPI
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'SoftSPI'
```

**Reply by robert-hh** [reply]
If speed does not matter, it's pretty easy  to implement SPI with Python. It's just a little bit of bit-banging.

**Reply by Meekdai** [reply]
I found that micropython v1.11 can use soft SPI by configured to `-1`. This is not introduced in the document. I found this hidden function after checking the source code, haha

```python
spi = machine.SPI(-1,sck=sck,miso=miso,mosi=mosi,phase=phase,polarity=polarity)
spi.write_readinto(data.to_bytes(1,'big'),rdata)
```

**Reply by rkompass** [reply]
[Here](https://docs.micropython.org/en/v1.9.2/wipy/library/machine.SPI.html#machine.SPI) for example it was documented:

> **classmachine.SPI(id, ...)**
Construct an SPI object on the given bus, id. Values of id depend on a particular port and its hardware. Values 0, 1, etc. are commonly used to select hardware SPI block #0, #1, etc. Value -1 can be used for bitbanging (software) implementation of SPI (if supported by a port).

This was changed in later versions. Nice that you found it yourself. How ?.. doesn't matter in the end :-)

**Reply by Meekdai** [reply]
I found the method in the following source code.
https://github.com/micropython/micropython/blob/6f75c4f3cd393131579db70cdf0b35d1fe5b95ab/extmod/machine_spi.c#L48-L63

**Reply by jimmo** [reply]
@Meekdai It _is_ in the documentation, you just have to look at the v1.11 documentation: https://docs.micropython.org/en/v1.11/library/machine.SPI.html#machine.SPI
We removed it from the docs when `machine.SoftSPI` was introduced, but the functionality remains for backwards-compatibility.

I strongly recommend updating to v1.21. There should be very few breaking changes (and possibly none at all).
